### PR TITLE
feat: watsonx support for traceloop

### PR DESCRIPTION
**Pull request by gyliu513**
Fixed https://github.com/traceloop/openllmetry/issues/339

### CODE REVIEW

**Comment by gyliu513** [comment]
Related with #340

File: packages/opentelemetry-instrumentation-watsonx/README.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# OpenTelemetry IBM Watsonx Instrumentation
+
+This library allows tracing VertexAI prompts and completions sent with the official [IBM Watson Machine Learning library](https://ibm.github.io/watson-machine-learning-sdk/).
+
+## Installation
+
+```bash
+pip install opentelemetry-instrumentation-watsonx
```
Comment (gyliu513) [inline packages/opentelemetry-instrumentation-watsonx/README.md:8]: @nirga does traceloop will help release opentelemetry-instrumentation-watsonx and let customer install via pip install?

**Comment by gyliu513** [comment]
@nirga can you help do a first round review, thanks!

File: packages/opentelemetry-instrumentation-watsonx/README.md (hunk 1 above)
Comment (nirga) [reply packages/opentelemetry-instrumentation-watsonx/README.md:8]: Yes :) We just need to update some stuff in the CI similar to these 2 together
https://github.com/traceloop/openllmetry/pull/285
https://github.com/traceloop/openllmetry/pull/316

Comment (gyliu513) [reply packages/opentelemetry-instrumentation-watsonx/README.md:8]: Awesome, thanks @nirga !

**Comment by gyliu513** [comment]
A simple test screenshot.

Screenshot 2024-01-19 at 12 07 38 PM

File: packages/opentelemetry-semantic-conventions-ai/opentelemetry/semconv/ai/__init__.py [hunk 2]
```diff
@@ -36,6 +36,14 @@ class SpanAttributes:
     # Deprecated
     TRACELOOP_CORRELATION_ID = "traceloop.correlation.id"
 
+    # Watson/genai LLM
+    LLM_DECODING_METHOD = "decoding_method"
+    LLM_RANDOM_SEED = "random_seed"
+    LLM_MAX_NEW_TOKENS = "max_new_tokens"
+    LLM_MIN_NEW_TOKENS = "min_new_tokens"
+    LLM_TOP_K = "top_k"
```
Comment (nirga) [inline packages/opentelemetry-semantic-conventions-ai/opentelemetry/semconv/ai/__init__.py:44]: There's already an `LLM_TOP_K`

File: packages/opentelemetry-semantic-conventions-ai/opentelemetry/semconv/ai/__init__.py [hunk 3]
```diff
@@ -36,6 +36,14 @@ class SpanAttributes:
     # Deprecated
     TRACELOOP_CORRELATION_ID = "traceloop.correlation.id"
 
+    # Watson/genai LLM
+    LLM_DECODING_METHOD = "decoding_method"
```
Comment (nirga) [inline packages/opentelemetry-semantic-conventions-ai/opentelemetry/semconv/ai/__init__.py:40]: I'd use `llm.watson.decoding_method` in line with what we're doing in https://github.com/open-telemetry/semantic-conventions/pull/639/files#diff-e352d0862431e100006cb429634845d199d25ad43c6d09c14e08267c7b73f5b4
(Same for others)

File: packages/opentelemetry-instrumentation-watsonx/README.md [hunk 4]
```diff
@@ -0,0 +1,21 @@
+# OpenTelemetry IBM Watsonx Instrumentation
+
+This library allows tracing VertexAI prompts and completions sent with the official [IBM Watson Machine Learning library](https://ibm.github.io/watson-machine-learning-sdk/).
```
Comment (nirga) [inline packages/opentelemetry-instrumentation-watsonx/README.md:3]:
```suggestion
This library allows tracing IBM Watsonx prompts and completions sent with the official [IBM Watson Machine Learning library](https://ibm.github.io/watson-machine-learning-sdk/).
```

**Review by nirga — COMMENTED** [review]
Looks good overall, made a few comments. 3 more things:

1. Can you share how this looks like on Traceloop / some other observability platform?
2. Do you plan on adding the CI/CD stuff or should I?
3. Can we add tests for this? If so how easy it is for me to actually run it against Watson?

**Comment by nirga** [comment]
> A simple test screenshot.
> 

Thanks! Can you show the other tabs maybe?

יש

**Comment by gyliu513** [comment]
Here are other screenshots, let me check why completion is empty.
Screenshot 2024-01-19 at 1 28 48 PM

Screenshot 2024-01-19 at 1 23 53 PM
Screenshot 2024-01-19 at 1 23 24 PM

**Comment by gyliu513** [comment]
@nirga 

> Can you share how this looks like on Traceloop / some other observability platform?

Does above screenshots looks good?

> Do you plan on adding the CI/CD stuff or should I?

I can help, but by checking https://github.com/traceloop/openllmetry/pull/285 and https://github.com/traceloop/openllmetry/pull/316 , not clear how those two PRs can help the release, do u have some other more related PRs?

> Can we add tests for this? If so how easy it is for me to actually run it against Watson?

I can add test, can the test be add by follow up PRs? Want to have this got merged and do some quick test in my team, comments?

File: packages/opentelemetry-instrumentation-watsonx/README.md (hunk 4 above)
Comment (gyliu513) [reply packages/opentelemetry-instrumentation-watsonx/README.md]: Done, thanks

File: packages/opentelemetry-semantic-conventions-ai/opentelemetry/semconv/ai/__init__.py (hunk 3 above)
Comment (gyliu513) [reply packages/opentelemetry-semantic-conventions-ai/opentelemetry/semconv/ai/__init__.py]: Done

File: packages/opentelemetry-semantic-conventions-ai/opentelemetry/semconv/ai/__init__.py (hunk 2 above)
Comment (gyliu513) [reply packages/opentelemetry-semantic-conventions-ai/opentelemetry/semconv/ai/__init__.py]: removed

**Review by gyliu513 — COMMENTED** [review]
Thanks @nirga for the quick response, comments are addressed.

**Comment by gyliu513** [comment]
> If so how easy it is for me to actually run it against Watson?

If you take a look at my test program https://github.com/gyliu513/openllmetry/blob/watsonx/packages/sample-app/sample_app/watsonx_generate.py#L21-L22 , only IAM_API_KEY and PROJECT_ID are required, you can start a free trial at https://dataplatform.cloud.ibm.com/registration/stepone?context=wx

**Comment by nirga** [comment]
It looks like the completion attribute is missing in the screenshot?

**Comment by nirga** [comment]
You were also missing the linter configuration so I've added it and there's some linting issues to fix.

**Comment by gyliu513** [comment]
Added completion

Screenshot 2024-01-19 at 6 43 58 PM

**Comment by gyliu513** [comment]
@nirga when I run `pip install .` under traceloop-sdk, always get following error:

```
INFO: pip is looking at multiple versions of opentelemetry-instrumentation-chromadb to determine which version is compatible with other requirements. This could take a while.
ERROR: Cannot install traceloop-sdk and traceloop-sdk==0.9.4 because these package versions have conflicting dependencies.

The conflict is caused by:
    traceloop-sdk 0.9.4 depends on opentelemetry-semantic-conventions-ai<0.0.16 and >=0.0.15
    opentelemetry-instrumentation-anthropic 0.9.4 depends on opentelemetry-semantic-conventions-ai<0.0.16 and >=0.0.15
    opentelemetry-instrumentation-bedrock 0.9.4 depends on opentelemetry-semantic-conventions-ai<0.0.16 and >=0.0.15
    opentelemetry-instrumentation-chromadb 0.9.4 depends on opentelemetry-semantic-conventions-ai<0.0.15 and >=0.0.14

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts
```

But I think I have updated the semantic convention ai to 0.15.0

```console
(py311) gyliu@guangyas-air traceloop-sdk % pip list | grep semantic
opentelemetry-semantic-conventions         0.43b0
opentelemetry-semantic-conventions-ai      0.0.15

[notice] A new release of pip is available: 23.2.1 -> 23.3.2
[notice] To update, run: pip install --upgrade pip
```

For the lint error, I check the CI log, but there is not too much detail for what is wrong with the code, how can I know the detail of the lint errors? Thanks

**Comment by gyliu513** [comment]
lint error only report the following, how do I know the exact error? Thanks

```

 >  NX   Running target lint for 17 projects failed

   Failed tasks:
   
   - opentelemetry-instrumentation-watsonx:lint

Error: Process completed with exit code 1.

```

**Comment by nirga** [comment]
> lint error only report the following, how do I know the exact error? Thanks
> 
> ```
> 
>  >  NX   Running target lint for 17 projects failed
> 
>    Failed tasks:
>    
>    - opentelemetry-instrumentation-watsonx:lint
> 
> Error: Process completed with exit code 1.
> ```

Just run it locally, from the root directory `nx run-many -t lint`. Here's the output I got:
```
ERROR  linting issues
       ./opentelemetry/instrumentation/watsonx/__init__.py:10:1: F401 'opentelemetry.trace.status.Status' imported but unused
       ./opentelemetry/instrumentation/watsonx/__init__.py:10:1: F401 'opentelemetry.trace.status.StatusCode' imported but unused
       ./opentelemetry/instrumentation/watsonx/__init__.py:100:1: W293 blank line contains whitespace
       ./opentelemetry/instrumentation/watsonx/__init__.py:109:121: E501 line too long (121 > 120 characters)
```

**Comment by gyliu513** [comment]
@nirga I fixed all of the lint errors you point out, but seems lint still failed. Can u enable the lint error can report what is wrong in CI so that I can check CI log and fix that?

I also run the nx locally, but failed.

```console
(py311) gyliu@guangyas-air opentelemetry-instrumentation-watsonx % npx nx run-many -t lint

node:internal/modules/cjs/loader:1147
  throw err;
  ^

Error: Cannot find module 'nx/presets/npm.json'
Require stack:
- /opt/homebrew/lib/node_modules/nx/src/config/nx-json.js
- /opt/homebrew/lib/node_modules/nx/src/config/configuration.js
- /opt/homebrew/lib/node_modules/nx/src/utils/package-manager.js
- /opt/homebrew/lib/node_modules/nx/bin/init-local.js
- /opt/homebrew/lib/node_modules/nx/bin/nx.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1144:15)
    at Function.resolve (node:internal/modules/helpers:187:19)
    at readNxJson (/opt/homebrew/lib/node_modules/nx/src/config/nx-json.js:13:48)
    at Object.<anonymous> (/opt/homebrew/lib/node_modules/nx/src/daemon/client/client.js:320:72)
    at Module._compile (node:internal/modules/cjs/loader:1376:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1435:10)
    at Module.load (node:internal/modules/cjs/loader:1207:32)
    at Module._load (node:internal/modules/cjs/loader:1023:12)
    at Module.require (node:internal/modules/cjs/loader:1235:19)
    at require (node:internal/modules/helpers:176:18) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/opt/homebrew/lib/node_modules/nx/src/config/nx-json.js',
    '/opt/homebrew/lib/node_modules/nx/src/config/configuration.js',
    '/opt/homebrew/lib/node_modules/nx/src/utils/package-manager.js',
    '/opt/homebrew/lib/node_modules/nx/bin/init-local.js',
    '/opt/homebrew/lib/node_modules/nx/bin/nx.js'
  ]
}

Node.js v21.2.0
```

**Review by nirga — APPROVED** [review]

**Pull request merged**